### PR TITLE
- Itroduce COAP

### PR DESCRIPTION
**Pull request by bitkis**
- Enable multiple routers
- Clean up
- Separate routers into different files
- Fix configuration parameter

### CODE REVIEW

**Comment by bitkis** [comment]
@Vitalius

Your registration of RTA handlers has been missed in merge, but I did not restore it intentionally: most likely, it needs to be moved. Please take a look and make corresponding changes.

File: include/router.h [hunk 1]
```diff
@@ -57,87 +55,90 @@ class RouterT
...
+                    );
+                }
+            );
+            char *errstr = NULL;
+            int err = r3_tree_compile(m_node, &errstr);
+
+            if (err != 0)
+                std::cout << "error: " << std::string(errstr) << std::endl;
+//          else
+//              r3_tree_dump(m_node, 0);
```
Comment (EDDragonWolf) [inline include/router.h]: Do we need it?

File: include/router.h [hunk 2]
```diff
@@ -57,87 +55,90 @@ class RouterT
...
 
-        return ret;
+        void addRouter(RouterT<In, Out>& r) { m_routers.push_front(std::move(r)); }
+
+    private:
+        R3Node *m_node;
+        std::forward_list<RouterT<In, Out>> m_routers;
+    };
+
+    RouterT(const std::string& prefix = "") : m_endpointPrefix(prefix) { }
```
Comment (EDDragonWolf) [inline include/router.h]: Maybe better `const std::string& prefix = std::string()`

File: src/main.cpp [hunk 3]
```diff
@@ -78,22 +82,24 @@ int main(int argc, const char** argv)
...
         const boost::property_tree::ptree& server_conf = config.get_child("server");
         sopts.http_address = server_conf.get<string>("http-address");
+        sopts.coap_address = server_conf.get<string>("coap-address");
         sopts.http_connection_timeout = server_conf.get<double>("http-connection-timeout");
         sopts.workers_count = server_conf.get<int>("workers-count");
         sopts.worker_queue_len = server_conf.get<int>("worker-queue-len");
 
-        // TODO configure router
-        graft::Router router;
-        graft::registerRTARequests(router);
```
Comment (EDDragonWolf) [inline src/main.cpp]: Looks like, you deleted registration for DAPI

**Pull request merged**